FlightJournal.Web/Aprs/AprsListener.cs

```diff
@@ -0,0 +1,165 @@
+﻿using System;
+using System.Linq;
+using Boerman.AprsClient;
+using Boerman.AprsClient.Enums;
+using Boerman.AprsClient.Models;
+using Boerman.Core.Spatial;
+using FlightJournal.Web.Logging;
+using Skyhop.FlightAnalysis;
+using Skyhop.FlightAnalysis.Models;
+
+namespace FlightJournal.Web.Aprs
+{
+    public interface IAprsListener : IDisposable
+    {
+        EventHandler<AircraftEvent> OnAircraftTakeoff { get; set; }
+        EventHandler<AircraftEvent> OnAircraftLanding { get; set; }
+
+        void Start();
+    }
+
+    public class AircraftEvent
+    {
+        public Aircraft Aircraft { get; }
+        public DateTime? Time { get; }
+
+        public AircraftEvent(Aircraft a, DateTime? t)
+        {
+            Aircraft = a;
+            Time = t?.ToLocalTime();
+        }
+    }
+    public class AprsListener : IAprsListener
+    {
+        private readonly IAircraftCatalog _catalog;
+        private static Listener _aprsClient;
+        private static readonly FlightContextFactory FlightContextFactory = new();
+
+        public EventHandler<AircraftEvent> OnAircraftTakeoff { get; set; }
+        public EventHandler<AircraftEvent> OnAircraftLanding { get; set; }
+
+
+        public AprsListener(IAircraftCatalog catalog)
+        {
+            _catalog = catalog;
+
+
+            FlightContextFactory.OnTakeoff += OnTakeoff;
+            FlightContextFactory.OnLanding += OnLanding;
+            FlightContextFactory.OnRadarContact += OnRadarContact;
+            FlightContextFactory.OnContextDispose += OnContactLost;
+
+            _aprsClient = new Listener(new Config
+            {
+                Callsign = "0",
+                Password = "-1",
+                Uri = "aprs.glidernet.org",
+                UseOgnAdditives = true,
+                Port = 14580, //10152: Full feed, 14580: Filtered
+                Filter = "r/55.84/10.63/220" //DK center 220km radius - multiple filters seem not to work ?
+            });
+
+            _aprsClient.DataReceived += OnAprsDataReceived;
+            _aprsClient.PacketReceived += OnAprsPacketReceived;
+        }
+
+        private void OnAprsPacketReceived(object sender, PacketReceivedEventArgs e)
+        {
+            switch (e.AprsMessage.DataType)
+            {
+                case DataType.Status:
+                    //                        Console.WriteLine(e.AprsMessage);
+                    break;
+                case DataType.PositionWithTimestampNoAprsMessaging:
+                case DataType.PositionWithTimestampWithAprsMessaging:
+                    //                        Console.WriteLine(e.AprsMessage);
+                    try
+                    {
+                        var positionUpdate = new Skyhop.FlightAnalysis.Models.PositionUpdate(
```

Comment:
Somewhere we need to mention Skyhop.FlightAnalysis as it has a GPL-v3

Reply:
Thanks, missed that.